### PR TITLE
Data sensitivity feature, fixes # 8

### LINKED ISSUE #8: Importance score for dataset training samples
Implement a way to assess the importance of an inidividual training datapoint to the performance of the model. This could be done by training the same model with a particular point included and then excluded from the training set, and computing the difference in performance scores on the test set.

### PR DESCRIPTION
**Pull request by Sidrah-Madiha**
This branch that I have merged fixes #8 ,
-  it  contains a helper function file "helper_for_senstivity_calculation.py" which has a function "calculate_senstivity" that returns L1 sensitivity of the dataset and list of performance metric for all parallel dataset
-  I have also added a function to display the most sensitive datapoint as well as the index of that data point in training dataset.
- I have tested it on the random forest classifier that I made for vehicles.csv dataset, please see test demo in jupyer notebook in folder: " Importance score for dataset training samples" >  vehicles_dataset_classifer_v1.ipynb file

@dzeber  I was not sure if the performance metric should be accuracy or score so I implemented it with score and commented out code that I used with accuracy . Please advise for the next steps to improve this fix.

### CODE REVIEW

File: dev/Sidrah-Madiha/Importance score for dataset training samples/helper_for_senstivity_calculation.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    inputs:
+    model_func : model for fitting training data
+    X_train : training data for model fitting
+    y_train : training data's target labels
+    X_test : testing data for model's accuracy calculation
+    y_test: testing data's target labels for model's accuracy calculation'''
+    model_func.fit(X_train, y_train)
+    # y_pred=model_func.predict(X_test) 
+    # return metrics.accuracy_score(y_test, y_pred)
+    # I am not sure if accuracy or score is record for senstivity calculation so I commented out accuracy code
```
Comment (dzeber) [inline dev/Sidrah-Madiha/Importance score for dataset training samples/helper_for_senstivity_calculation.py]: I'm pretty sure these do the same think. I thing `model_func.score()` is just a shortcut.

File: dev/Sidrah-Madiha/Importance score for dataset training samples/vehicles_dataset_classifer_v1.ipynb [hunk 2]
```diff
@@ -0,0 +1,826 @@
...
+      "0.11176470588235288\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(\"priniting model performance score for all parallel datasets\", list_scores)\n",
```
Comment (dzeber) [inline dev/Sidrah-Madiha/Importance score for dataset training samples/vehicles_dataset_classifer_v1.ipynb]: Could you show a visualization instead of the full list?

File: dev/Sidrah-Madiha/Importance score for dataset training samples/helper_for_senstivity_calculation.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+def calculate_senstivity(model_func, X_train, y_train, X_test, y_test):
+    ''' calculates L1 senstivity of dataset and returns maximum senstivity as well as list of all accuracies of parallel datasets'''
+    full_dataset_score = model_performance_evaluater_score(model_func,X_train, y_train, X_test, y_test)
+    sensitivity = 0
+    score_all_db = []
+    list_train_data_label =create_all_parallel_dbs(X_train, y_train) #return list containing tuples of training data and label
+    #print(len(list_train_data_label ))
+    for parallel_train, parallel_label in list_train_data_label:
+        parallel_dataset_score =model_performance_evaluater_score(model_func,parallel_train, parallel_label, X_test, y_test)
```
Comment (dzeber) [inline dev/Sidrah-Madiha/Importance score for dataset training samples/helper_for_senstivity_calculation.py]: Rather than precomputing all the parallel DBs, why not call `create_parallel_db()` inside this loop and compute the score right away for each. This will probably be easier for larger datasets.

File: dev/Sidrah-Madiha/Importance score for dataset training samples/helper_for_senstivity_calculation.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        dataset_distance = np.abs(full_dataset_score - parallel_dataset_score ) # L1 senstivity
```
Comment (dzeber) [inline dev/Sidrah-Madiha/Importance score for dataset training samples/helper_for_senstivity_calculation.py]: I'm thinking it would be interesting compute the distance without the absolute value. It would probably be valuable to know if certain training samples actually improve the model fit when being left out.

**Comment by mlopatka** [comment]
@Sidrah-Madiha please let us know if you intend to continue with this PR or if this work has been integrated to another PR and this one can be closed out.

**Comment by Sidrah-Madiha** [comment]
@mlopatka , yes I am working on fixing it today.

**Comment by Sidrah-Madiha** [comment]
@mlopatka please accept my PR for this issue, I have fied all change requests.
- reformatted code for helper file with black.
- showed visualization of full list of scores for each parallel db. Please see image attached
sens2
- fixed this issue _(Rather than precomputing all the parallel DBs, why not call create_parallel_db() inside this loop and compute the score right away for each. This will probably be easier for larger datasets.)_ please see attached image for changes for
sens1
 code refactoring
- removed absolute function from senstivity distance difference calculation. (showed in above iage aas well)

**Comment by Sidrah-Madiha** [comment]
Hi @dzeber, please merge this PR, I have fixed all change requests

**Comment by Sidrah-Madiha** [comment]
Hi @dzeber I have further improved this fix, I have added a scoring parameter with default "accuracy_score in function "model_performance_evaluater_score"
I have fixed relative path issue in notebook
I have added an indicator in the visualization to indicate the most sensitive datapoint, fixed labels to have better description. 
sensitive

Please review
